### DISCUSSION TITLE
it is conceptually mistake to use log loss for performance comparison

### DISCUSSION THREAD

**Discussion by Sandy4321** [manujosephv/pytorch_tabular#255] [Q&A]
**Describe the bug**
it is conceptually mistake to use log loss for performance comparison 
 since 
1
lightgbm as all gmb models
 (  maybe except https://stanfordmlgroup.github.io/projects/ngboost/ NGBoost: Natural Gradient Boosting for Probabilistic Prediction
Tony Duan*, Anand Avati*, Daisy Yi Ding, Sanjay Basu, Andrew Ng, Alejandro Schuler  ) 
are not good in prediction probabilities
2
then same should be carefully tested for deep NN

3
may you change provement that your code is performing , for example 
https://github.com/manujosephv/pytorch_tabular/blob/main/examples/PyTorch%20Tabular%20with%20Bank%20Marketing%20Dataset.ipynb
to compare actual confusion matrices pls 
I did some attempt 
preds_GATE_Full  = tabular_model.predict(test)
but what is the way to calculate ground truth predictions from your models  
![image](https://github.com/manujosephv/pytorch_tabular/assets/11426119/1b7225a9-92ed-4c2f-a2ff-2074d9bb4c3a)

4
where is description of used algorithms , especially for categorical data values preprocessing ?

**Comment by Sandy4321** [comment]
ok , i found how to calculate performance 
![image](https://github.com/manujosephv/pytorch_tabular/assets/11426119/9456618f-0e9e-4ec0-82c3-5dbb210f6b90)

**Comment by Sandy4321** [comment]
so
confusion_matrix for GATE_Full
array([[7765,  220],
       [ 653,  405]], dtype=int64)
confusion_matrix for lightgbm
array([[7684,  301],
       [ 528,  530]], dtype=int64)

when modle training time much less for lightgbm ( I used cpu runs)
what is about inference time?
what is about ram and cpu/gpu resources use?

**Comment by manujosephv** [comment]
@Sandy4321 Thank you for checkout the library... And I think you might be reading too much into the tutorial.

Let me try to address your points:

1. I don't think comparing logloss is a problem. Even GBMs produce confidence scores and we convert them based on the 0.5 cutoff. If you end up using the score for ranking or deciding on custom thresholds etc, you should be looking at the goodness of your confidence scores. But, either way, the notebook prints Accuracy and LogLoss
2. the notebook is purely a tutorial and it does not make the claim that all those methods are better or worse than LGBM. It is merely a tutorial to use different models. Also, making the case that a model is better or worse (in general) can't be made using a single dataset. And also, that is what research papers are for. If I were to make that kind of study and claim, it will definitely be in an arxiv paper to say the least.
3. If this question is regarding GATE model, then the arxiv paper will tell you all you need to know.
4. PyTorch Tabular uses entity embeddings by default to handle categorical features

I hope I have answered your questions. Let me know if you have more..

**Comment by Sandy4321** [comment]
thanks for soon replay 
1
I don't think comparing logloss is a problem. Even GBMs produce confidence scores and we convert them based on the 0.5 cutoff. If you end up using the score for ranking or deciding on custom thresholds etc, you should be looking at the goodness of your confidence scores. But, either way, the notebook prints Accuracy and LogLoss

theoretically it is obvious GBM can not calculate  confidence scores
did you checked reference I gave you 
https://stanfordmlgroup.github.io/projects/ngboost/ NGBoost: Natural Gradient Boosting for Probabilistic Prediction
Tony Duan*, Anand Avati*, Daisy Yi Ding, Sanjay Basu, Andrew Ng, Alejandro Schuler 
2
the notebook is purely a tutorial and it does not make the claim that all those methods are better or worse than LGBM. It is merely a tutorial to use different models. Also, making the case that a model is better or worse (in general) can't be made using a single dataset. And also, that is what research papers are for. If I were to make that kind of study and claim, it will definitely be in an arxiv paper to say the least.

do you have some evidence that PyTorch Tabular  is better then generally used ML algorithms ?

3
If this question is regarding GATE model, then the arxiv paper will tell you all you need to know.
what is link pls
4
PyTorch Tabular uses entity embeddings by default to handle categorical features
it is crucial part of our code - do you have detailed description what  math / algorithms do you suggest to use 


thanks for taking care

**Reply by manujosephv** [reply]
1. 
I'm aware about NGBoost and the difference between that and more popular GBMs like XGBoost. In fact, I had written about all of these 3 years before in a blog series [here](https://deep-and-shallow.com/2020/02/02/the-gradient-boosters-i-the-math-heavy-primer-to-gradient-boosting-algorithm/) 

I'm not entirely clear why you say "GBM's can't calculate confidence scores". Is it from the point of view that the confidence scores aren't true probabilities?

2. 
PyTorch Tabular is not an algorithm, but a library which implements some algorithms. So, the question should be, is Pytorch Tabular better than scikit-Learn. And saying PyTorch Tabular is worse or better than Sci-kit Learn is like saying if python is better than R. Both are tools to do a job and PyTorch Tabular just adds a few more tools to your toolbelt.
And lastly, the debate whether DL models are better than ML models has been going on for a long time.  But if you keep No Free Lunch Theorem in mind, and choose the right tool for the job, there is no need for a debate after all.

3.
Papers for all the models implemented in PyTorchTabular is [here](https://github.com/manujosephv/pytorch_tabular/#available-models)

4.
Entity Embeddings are really similar to Word Embeddings, and it is a really common way to handle categorical features. You can read up more in countless blogs and papers. One eg, [here](https://arxiv.org/abs/1604.06737)
